neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+  1) It would require a massive migration of trie.
+  2) We would need to maintain the old and the new trie structure forever.
+* Changing the storage structure by having the storage key to have the format of `account_id.node_hash`. This structure would make it much easier to split the trie on storage level because the children shards are simple sub-ranges of the parent shard. Unfortunately we found that the migration would not be feasible. 
+* Changing the storage structure by having the key format as only node_hash and dropping the ShardUId prefix. This is a feasible approach but it adds complexity to the garbage collection and data deletion, specially when nodes would start tracking only one shard. We opted in for the much simpler one by using the existing scheme of prefixing storage entries by shard uid. 
+
+#### Other considerations
+
+* Dynamic Resharding - we have decided to not implement the full dynamic resharding at this time. Instead we hardcode the shard layout and schedule it manually. The reasons are as follows:
+  * We prefer incremental process of introducing resharding to make sure that it is robust and reliable, as well as give the community the time to adjust. 
+  * Each resharding increases the potential total load on the system. We don't want to allow it to grow until full sharding is in place and we can handle that increase. 
+* Extended shard layout adjustments - we have decided to only implement shard splitting and not implement any other operations. The reasons are as follows:
+  * In this iteration we only want to perform splitting.
+  * The extended adjustments are currently not justified. Both merging and boundary moving may be useful in the future when the traffic patterns change and some shard become underutilized. In the nearest future we only predict needing to reduce the size of the heaviest shards. 
+
+### What is the impact of not doing this?
+
+We need resharding in order to scale up the system. Without resharding eventually shards would grow so big (in either storage or cpu usage) that a single node would not be able to handle it. Additionally, this clears up the path to implement in-memory tries as we need to store the whole trie structure in limited RAM. In the future smaller shard size would lead to faster syncing of shard data when nodes start tracking just one shard.
+
+## Integration with State Sync
+
+There are two known issues in the integration of resharding and state sync:
+
+* When syncing the state for the first epoch where the new shard layout is used. In this case the node would need to apply the last block of the previous epoch. It cannot be done on the children shard as on chain the block was applied on the parent shards and the trie related gas costs would be different.
+* When generating proofs for incoming receipts. The proof for each of the children shards contains only the receipts of the shard but it's generated on the parent shard layout and so may not be verified. 
```

Comment:
@karim-en will this have an impact on the bridge?

Reply:
It looks like yes, but I don't understand how this could impact the Prover logic, so I would like to see the structure of the newly generated proof.
Also, it would be good if @walnut-the-cat had a look into our proof verification code and gave us some feedback.
https://github.com/aurora-is-near/rainbow-bridge/blob/bf7cc1e81ee4a4fad3e74e2c599b0b6cf3757f4d/contracts/eth/nearprover/contracts/NearProver.sol#L28

Reply:
@birchmd and @karim-en I had a quick look but I completely lack context. Perhaps it would be easier to perform an end to end test of the bridge and resharding on localnet or other small network? A simple setup to trigger resharding would be as follows:
* prepare custom binary with resharding stabilized e.g. like so https://github.com/near/nearcore/commit/fdcfedf01d296358aa1146ade6b7e723a31ae0cb 
* run 1.36 for an epoch or two
* stop the node(s)
* run the custom binary for at least three epochs
* assuming the binary upgrade happens in epoch T, the resharding will happen in T+1, and the network will switch to the new shard layout in the first block of T+2

Please let me know if you need any help at all and we can look into it together.